File: typescript/src/data-structures/linked_list.ts
/**
 * Singly Linked List
 * 
 * A linear data structure where elements are not stored in contiguous memory locations.
 * 
 * Time Complexity:
 * - Access: O(n)
 * - Search: O(n)
 * - Insertion (at head): O(1)
 * - Deletion (at head): O(1)
 * 
 * Use when:
 * - Frequent insertions and deletions are required.
 * - Dynamic size is needed without pre-allocation.
 * 
 * Avoid when:
 * - Random access is required.
 * - Memory overhead of pointers is a concern.
 */

class LinkedListNode<T> {
  constructor(public value: T, public next: LinkedListNode<T> | null = null) {}
}

export class LinkedList<T> {
  private head: LinkedListNode<T> | null = null;
  private size: number = 0;

  /**
   * Add an element to the front of the list.
   */
  prepend(value: T): void {
    this.head = new LinkedListNode(value, this.head);
    this.size++;
  }

  /**
   * Add an element to the end of the list.
   */
  append(value: T): void {
    const newNode = new LinkedListNode(value);
    if (!this.head) {
      this.head = newNode;
    } else {
      let current = this.head;
      while (current.next) {
        current = current.next;
      }
      current.next = newNode;
    }
    this.size++;
  }

  /**
   * Remove the first occurrence of a value.
   */
  remove(value: T): boolean {
    if (!this.head) return false;

    if (this.head.value === value) {
      this.head = this.head.next;
      this.size--;
      return true;
    }

    let current = this.head;
    while (current.next && current.next.value !== value) {
      current = current.next;
    }

    if (current.next) {
      current.next = current.next.next;
      this.size--;
      return true;
    }

    return false;
  }

  toArray(): T[] {
    const arr: T[] = [];
    let current = this.head;
    while (current) {
      arr.push(current.value);
      current = current.next;
    }
    return arr;
  }

  getSize(): number {
    return this.size;
  }
}

File: c/data_structures/linked_list.c
/**
 * Singly Linked List
 * Time Complexity:
 * - Access/Search: O(n)
 * - Insertion/Deletion (Head): O(1)
 * Space Complexity: O(n)
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

typedef struct Node {
    int value;
    struct Node* next;
} Node;

typedef struct LinkedList {
    Node* head;
    int size;
} LinkedList;

LinkedList* createList() {
    LinkedList* list = (LinkedList*)malloc(sizeof(LinkedList));
    list->head = NULL;
    list->size = 0;
    return list;
}

void append(LinkedList* list, int value) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->value = value;
    newNode->next = NULL;

    if (list->head == NULL) {
        list->head = newNode;
    } else {
        Node* current = list->head;
        while (current->next != NULL) {
            current = current->next;
        }
        current->next = newNode;
    }
    list->size++;
}

void freeList(LinkedList* list) {
    Node* current = list->head;
    while (current != NULL) {
        Node* next = current->next;
        free(current);
        current = next;
    }
    free(list);
}

int main() {
    LinkedList* ll = createList();
    append(ll, 1);
    append(ll, 2);
    append(ll, 3);

    assert(ll->size == 3);
    assert(ll->head->value == 1);
    assert(ll->head->next->value == 2);
    assert(ll->head->next->next->value == 3);

    printf("C Linked List: All tests passed!\n");
    freeList(ll);
    return 0;
}
